urlc8dbc7ce6e_dlubarov_guava_tgz-pJ8-MainJ8/src/d/TypeDef.java

```diff
@@ -0,0 +1,139 @@
+package d;
+
+import java.util.*;
+
+import util.StringUtils;
+
+import common.*;
+import d.ty.ConcreteType;
+import d.ty.desc.TypeDesc;
+import d.ty.sup.*;
+
+public class TypeDef {
+    public final RawType desc;
+    public Variance[] genericVariances;
+
+    public final BaseObject[] staticFields;
+    public final int numInstanceFields;
+
+    public ConcreteMethodDef[] staticMethods;
+    public MethodDef[] instanceMethods;
+
+    public final IdentityHashMap<String, Integer> virtualFieldTable;
+
+    public Map<RawMethod, RawMethod> virtualMethodDescTable;
+    public final Map<MethodDef, ConcreteMethodDef> virtualMethodTable;
+
+    public Map<RawType, TypeDesc[]> superGenericDescs;
+    public final Map<TypeDef, SuperType[]> superGenerics;
+
+    // 0 = not initialized, 1 = in progress, 2 = initialized
+    public int initializationStatus = 0;
+
+    private ConcreteType typeNoGenerics = null;
+
+    public TypeDef(RawType desc,
+            Variance[] genericVariances,
+            int numStaticFields,
+            String[] allInstanceFields,
+            ConcreteMethodDef[] staticMethods,
+            MethodDef[] instanceMethods,
+            Map<RawMethod, RawMethod> virtualMethodDescTable,
+            Map<RawType, TypeDesc[]> superGenericDescs) {
+        this.desc = desc;
+        this.genericVariances = genericVariances;
+
+        this.staticFields = new BaseObject[numStaticFields];
+        numInstanceFields = allInstanceFields.length;
+
+        this.staticMethods = staticMethods;
+        this.instanceMethods = instanceMethods;
+
+        this.virtualFieldTable = new IdentityHashMap<String, Integer>();
+        for (int i = 0; i < numInstanceFields; ++i)
+            virtualFieldTable.put(allInstanceFields[i].intern(), i);
+
+        this.virtualMethodDescTable = virtualMethodDescTable;
+        virtualMethodTable = new HashMap<MethodDef, ConcreteMethodDef>();
+
+        this.superGenericDescs = superGenericDescs;
+        superGenerics = new HashMap<TypeDef, SuperType[]>();
+
+        if (genericVariances == null || genericVariances.length == 0)
+            typeNoGenerics = new ConcreteType(this);
+
+        God.newType(this);
+    }
+
+    public void link() {
+        for (MethodDef method : staticMethods)
+            method.link();
+        for (MethodDef method : instanceMethods)
+            method.link();
+
+        // Unless this is an abstract type, populate the virtual method table.
+        if (virtualMethodDescTable != null)
+            for (RawMethod aDesc : virtualMethodDescTable.keySet()) {
+                RawMethod bDesc = virtualMethodDescTable.get(aDesc);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.